Add license directive

boomanaiden154 opened the pull request:
This patch adds in a license directive to get the ball rolling on adding in license information about packages to spack. I'm primarily interested in just adding license into spack, but this would also help with other efforts that people are interested in such as adding license information to the ASP solve for concretization to make sure licenses are compatible.

boomanaiden154 committed:
Add license directive

boomanaiden154 commented:
I'm not exactly sure what else needs to be done here, but this definitely isn't complete. I tried looking at some other directive commits/PRs for what to do but they all seemed to be quite varied. Here are the following two things that I think are currently necessary and I would very much appreciate hearing from others what else they think is necessary:
* [ ] Documentation on the new directive
* [ ] Outputting license information through a CLI command, probably as part of the spec.

alecbcs commented:
@boomanaiden154 This is cool! 
1. `spack info` could be a good candidate for displaying a package's stored license info.
2. We should look to add the license directive into the templates created by `spack create`.

boomanaiden154 committed:
Add license directive to spack create template

boomanaiden154 committed:
Add license printing to spack info

boomanaiden154 committed:
Add documentation on license directive

boomanaiden154 commented:
Need to do formatting at some point (don't have the style stuff setup locally), but the TODOs should be done. SPDX validation is going to be more complicated than I thought as clauses like `MIT OR BSD-3-Clause` are valid, so I'd need to do a lot more parsing logic if we want to support the full specification.

I think what's remaining is the following:
* There are some cases where projects will be subject to multiple licenses at the same time and other cases where projects will be available under multiple licenses at the same time (like Spack). Probably not the most important detail for an early implementation. I'm thinking parallel licenses (like what Spack has) can be modeled using multiple license directives and packages subject to the restrictions of multiple licenses will be a single directive. I think that would be easiest if anyone wants to plug it into ASP, but I'm not totally sure. This would also mean an identifier parser would be pretty easy to write.
* We'll need to have a custom identifier for projects that are all rights reserved/not released under an open source license. I'm thinking we could adopt NPM's `UNLICENSED`, but am open to suggestions.
* Maybe a disclaimer somewhere that the license annotations are best effort and not guaranteed to be correct?

boomanaiden154 commented:
@spackbot fix style

boomanaiden154 committed:
[@spackbot] updating style on behalf of boomanaiden154

boomanaiden154 committed:
Fix unit test

alecbcs commented:
> ...I'm thinking parallel licenses (like what Spack has) can be modeled using multiple license directives and packages subject to the restrictions of multiple licenses will be a single directive.

Instead of multiple license directives what if we supported multiple licenses in a single directive similar to what we currently do with `maintainers("someone", "someone-else")`? Is there additional overhead here that makes multiple directives better?

alecbcs commented:
This looks good to me. Going to ask @tgamblin, @alalazo, @becker33 to review and have the final word before we merge.

alecbcs commented:
@boomanaiden154 from a discussion with @tgamblin. It would be good if could add `when=` support to this to handle packages that change licenses over time. Perhaps something like,

```
license("UNLICENSED", when="@0.0.1:1")
license("MIT", when="@1:")
```

boomanaiden154 commented:
> @boomanaiden154 from a discussion with @tgamblin. It would be good if could add `when=` support to this to handle packages that change licenses over time. Perhaps something like,
> 
> ```
> license("UNLICENSED", when="@0.0.1:1")
> license("MIT", when="@1:")
> ```

It should have support for when conditions. I'm not completely sure I've implemented it correctly, but it should be there.

alecbcs commented:
Oh you're right. Would it be possible to bring in the when data into `spack info`? Perhaps with a table similar to variants?

alecbcs commented on lib/spack/spack/test/directives.py:
```diff
@@ -89,6 +89,12 @@ def test_maintainer_directive(config, mock_packages, package_name, expected_main
     assert pkg_cls.maintainers == expected_maintainers
 
 
+@pytest.mark.parametrize("package_name,expected_licenses", [("licenses-1", ["MIT", "Apache-2.0"])])
+def test_license_directive(config, mock_packages, package_name, expected_licenses):
+    pkg_cls = spack.repo.PATH.get_pkg_class(package_name)
+    assert expected_licenses == list(pkg_cls.licenses.keys())
```
Let's also add a check here for `when=` to test it's being set as we'd expect. Otherwise this is looking really good.

tgamblin commented on lib/spack/spack/directives.py:
```diff
@@ -863,6 +864,29 @@ def _execute_maintainer(pkg):
+        if not when_spec:
+            return
+
+        when_spec_list = pkg.licenses.setdefault(license_identifier, [])
+        when_spec_list.append(when_spec)
```
I'm in the middle of a refactor on how we store data on packages.  Short story: all dictionaries on the package are going to be keyed by `when` spec.  So, can you invert this one?  Basically instead of storing a dict like:

```
licenses = { license_id: [when_spec, ...] }
```

Do this instead:

```
licenses = { when_spec: license_id }
```

I think the directive should also raise an error if you try to specify different license ids for the same `when` spec, or, really, if any of the `when` specs overlap.  You can easily check the former as the directive is called (basically if there is already a license id and it's different, complain).  The latter probably needs to be done after all the meta class stuff executes... you can use `Spec.intersects()` to do it.  e.g. `foo+bar` and `foo~bar` do not intersect but `foo` and `foo+bar` do (b/c there are concrete specs that could match both).

boomanaiden154 replied:
I've implemented checking for overlapping specs by just calling `Spec.intersect` within the `_execute_license` function. It seems to cover the cases that I can think of. I guess that's only run as the meta class is executing though?

boomanaiden154 committed:
Reverse dict, add testing for when spec

boomanaiden154 commented:
> Oh you're right. Would it be possible to bring in the when data into `spack info`? Perhaps with a table similar to variants?

I've added the additional information into `spack info` in a similar format to how versions are presented currently. The table formatting looks fairly involved and maybe a little bit overkill for data presentation that only needs two columns.

boomanaiden154 committed:
Add ability to print license when specs

boomanaiden154 committed:
Add duplicate license handling

boomanaiden154 committed:
Add support for detecting overlapping license when specs

tgamblin commented on lib/spack/docs/packaging_guide.rst:
```diff
@@ -6606,3 +6606,30 @@ To achieve backward compatibility with the single-class format Spack creates in
+certain build configurations that include files that are licensed differently.
+To account for this, you can specify when licenses should be applied. For
+example, to specify that a specific license identifier should only apply
+to versions before 1.5, you could write the following directive:
```
The version range below is actually versions *up to and including* 1.5

tgamblin commented on lib/spack/spack/cmd/create.py:
```diff
@@ -62,6 +62,9 @@ class {class_name}({base_class_name}):
     # maintainers("github_user1", "github_user2")
 
+    # FIXME: Add the SPDX identifier of the project's license below.
+    license("UNLICENSED")
```
Can you make this `UNKNOWN`

tgamblin reviewed (changes requested):
This is looking great -- do you mind making a couple more minor changes?  I think this is nearly ready.

A couple things we should think about doing as follow-on work:
1. Include the license data in concrete Specs (probably something we can do at end of concretization).  I might integrate that with #32312.
2. Show license data in `spack spec` (and `spack find`, etc. -- basically update `Spec.tree()`)
3. Stretch goal: think about how to integrate licenses with dependency solving -- e.g. user could say things like "get me a build that's fully permissive" or "only install Apache-compatible software"

boomanaiden154 committed:
Address reviewer feedback

boomanaiden154 commented:
Those follow-up mini-projects definitely sound like they could be pretty interesting, particularly getting solving by license working (which I guess has a prerequisite on the other two). I'd be interested in getting solving by license working, but I'm not sure what my availability will be like as I start school on the 27th and am working on at least working on other projects in parallel.

Another project that needs to be done afterwards is to actually add license information to existing packages. We'll probably need to come up with some strategy to do this as it will end up creating changes across many packages that we will probably want to have reviewed by maintainers when they are specified. Even if I only end up working on the set of packages that I care about, that's still 5000-6000 packages or so. I'm thinking something like the following:

1. When a package has maintainers open up a separate PR per package updating license information, ideally including necessary context such as links to license change decisions, license in the source history, etc. This allows maintainers to comment/approve, validating everything and keeps context fairly clean.
2. For packages without maintainers, some sort of chunking while ideally keeping similar context to above. Maybe 10 or so packages per PR?

Ideally the community would also perform a decent amount of the license annotations, but I'm sure there are quite a few packages that aren't too actively maintained. I'll have to see how many packages I can get through (probably with the assistance of some automated tooling).

And then at some point I think it might be reasonable to require packages to have license information.

boomanaiden154 committed:
Update test related to last changes

tgamblin commented on lib/spack/spack/cmd/info.py:
```diff
@@ -71,6 +71,10 @@ def variant(s):
 
 
+def license(s):
+    return spack.spec.version_color + s + plain_format
```
```suggestion
    return spack.spec.VERSION_COLOR + s + plain_format
```

This changed on `develop`, which is why CI is failing.

boomanaiden154 replied:
Rebased and fixed. Thanks for suggesting the fix!

tgamblin commented:
> Another project that needs to be done afterwards is to actually add license information to existing packages.

Heh yes I'm getting ahead of myself 😅 

I'm fine with adding license information gradually and allowing maintainers to approve individually if you want.  Note that package maintainers, at least currently, provide advisory reviews -- i.e. they don't *have* to review all changes to their packages (though we do like them too).  So, I don't think we *have* to do one PR per package even with maintainers.

I think for some packages, like ones with valid PyPI links, you could automate getting the license information from, say, PyPI.  for others yes it's harder.  We can certainly encourage people on slack and other places to add license information.  We could also get `spackbot` to nudge people to provide it.

boomanaiden154 committed:
Fix test failures after rebase

boomanaiden154 commented:
> I'm fine with adding license information gradually and allowing maintainers to approve individually if you want. Note that package maintainers, at least currently, provide advisory reviews -- i.e. they don't _have_ to review all changes to their packages (though we do like them too). So, I don't think we _have_ to do one PR per package even with maintainers.

Sounds good. I'll probably chunk it a bit more then so that we don't end up with a couple thousand PRs.

> I think for some packages, like ones with valid PyPI links, you could automate getting the license information from, say, PyPI. for others yes it's harder. We can certainly encourage people on slack and other places to add license information. We could also get `spackbot` to nudge people to provide it.

PyPI is definitely a good suggestion. From a quick sample of maybe ~20 functions, it seems like maybe 25-50% of projects are also in traditional Linux distros, so I'll probably grab license information from there too. That'll cover at least some amount of the spack repository. THe only potential issue I see with this is that it doesn't capture a lot of license subtleties that could come up (vendored code under a different license, license changes between version (although this is feasible with more tooling in some cases), probably more situations), but I guess the focus at first should probably be to have some license information rather than none and I would imagine that more active projects that receive more attention would be the ones doing things like license changes or experience more subtleties than others (but I could be very wrong here).

boomanaiden154 commented:
@spackbot run pipeline

boomanaiden154 commented:
It looks like there's one (transient?) unit test failure in the CI related to another mock package somehow gaining license information that it shouldn't have. I'm seeing one similar failure locally when using `xdist` but can't reproduce when running the test/test-suite individually (`lib/spack/spack/test/directory_layout.py::test_yaml_directory_layout_parameters`), even with `xdist`. What's incredibly odd is that the failure happens deterministically when running all the unit tests with `xdist`. The test that is failing on CI is `lib/spack/spack/test/spec_yaml.py::test_save_dependency_spec_jsons_subset `. I did a little bit of digging and couldn't see anything obvious.

Any pointers on where I should start digging for the underlying issue?

tgamblin committed:
Merge branch 'develop' into license-directive

boomanaiden154 committed:
Merge branch 'develop' into license-directive

boomanaiden154 committed:
Fix transient failures in other mock packages

boomanaiden154 commented:
So apparently I don't really understand the `@directive` decorator as running a line like the following:
```
license_directive = spack.directives.license("MIT", "+foo")
```
causes transient failures when using `xdist` while loading other mock packages. I've refactored to use a pattern similar to what was used with the `version` annotation and split the `_execute_license` function out of the `license` directive scope so that we can unit test it individually and everything seems to be fixed now.

I think this PR should be good to go

boomanaiden154 commented:
I've also done some preliminary experiments on automatically tagging license information. I probably need to hack on the Pypi section a bit more, but currently, I'm only able to automatically tag about 2000/7500 packages. There are about 1300 python packages that I'm not currently tagging automatically due to Pypi's license field being arbitrary text and often not containing anything close to a SPDX identifier. They have some license classifiers that are present on about 80% of packages but don't correspond to a specific license (e.g., the classifier might say BSD without specifying which BSD license), which makes them not particularly useful (at least for most licenses).
